0001-scripting.md

```diff
@@ -0,0 +1,143 @@
+
+### Runner
+
+The runner part of the language driver is attached to the game's runtime. They load the built script and run them in their proper context depending on their type. For example, if a script is a `System`, then the runner adds a new system in the dispatcher that will handle the running of that specific script.
+
+In practice, the runner will also have to handle parallelism for languages that do not natively support it. See the Lua implementation details below for an example of such a situation.
+
+Additionally, the runner needs to enforce safety rules. Part of it can be enforced by restricting the available APIs at build time. 
+
+The first rule is isolation: in order to enforce mutability rules, `System`s must not be able to communicate with each other directly. How this is done in practice depends on the scripting language (see Lua below for an example).
+
+The second rule is enforcing ECS lifetime guarantees. We need to make sure the scripts cannot use the references the dispatcher is providing them in a dispatcher stage where it is unsafe. We found multiple solutions to guarantee this. The first one is to ban scripts from handling their own threading. This will make sure execution will stop at the end of the `run` method, guaranteeing the references can only be used when the dispatcher knows it is safe. However, banning threading entirely might be too much of a constraint. We do not believe so, but we designed alternatives anyway because we love you. To keep threading and still enforce lifetimes, we can either use a runtime lifetime check called "reference gates", or provide limited threading.
+
+##### Reference gates
+
+Reference gates act as a lifetime runtime check. When the dispatcher gives a reference to the language driver, it wraps it into a box with a gate. The gate is controlled by the dispatcher. As long as it is open, the reference can be used freely (it is however abstracted away, the script cannot copy it). But as soon as it is closed, it will panic if used.
+
+This has the obvious drawback of totally happening at runtime, costing a tiny bit and potentially panicking in convoluted situations.
+
+##### Limited threading
+
+Another alternative would be to provide limited managed threading, in a way similar to JavaScript WebWorkers: jobs running in a parallel thread without shared memory, communicating with memory ownership transfers. Those transfers would be carefully monitored to disallow the passing of references.
+
+A drawback of this approach is that communication between threads can not be fast because it is message-based. It might also involve a complicated implementation.
+
+
+Anyway, keep in mind that what approach is chosen can be different for each language, as different languages give us more or less flexibility (Rust as a scripting language for example already enforces all of this out of the box).
+
+## Hotswapping
```

Comment:
Does this mean that script system(s) are run in a separate `World`(?). I don't remember where I read it but since there is no way to add or remove systems after constructing a `World` it was recommended to keep dynamic systems in a separate world to avoid needing to reconstruct everything everytime something is reloaded.

https://github.com/Marwes/shred-example

https://github.com/Marwes/shred-example/blob/9f6d6a62f0b4cf7f55c3e5e9ca851c76c8ec9dea/src/gluon_system.rs#L798-L800

Reply:
The architecture Specs will have in the future will differ from the current quite a bit. It's already in design ;)

Reply:
Regarding your comment, I think you're mixing up `Dispatcher` and `World`. Both are very likely going to change (the latter one also renamed to `SystemGraph` which will not need any of those workarounds).